### ISSUE TITLE
Refactor maxNativeZoom into GridLayer

### ISSUE THREAD

**Issue by IvanSanchez** [Leaflet/Leaflet#5316]
So after pondering about https://gitlab.com/IvanSanchez/Leaflet.GridLayer.GoogleMutant/issues/16 , I think that it would be a good idea to refactor the `maxNativeZoom` (and `minNativeZoom`) out of `TileLayer` and into `GridLayer`.

I *think* that adding a tiny bit of extra logic in `_setView` [around here](https://github.com/Leaflet/Leaflet/blob/release-v1.0.3/src/layer/tile/GridLayer.js#L494) should do the trick. Just **clamp the calculated `tileZoom`** value to the `min`/`maxNativeZoom` if they are not `undefined`.

Doing so *should* skip the call to `_updateLevels()` when zooming past the min/max "native" range, and the CSS scale transforms for the levels container should do the rest.

As a bonus, this would also be a solution for https://github.com/Leaflet/Leaflet/pull/3811 .

The algorithm looks simple. If it works, then the rest is just bits of clerical work:
* [x] Create/modify a debug example such as `debug/map/grid.html` to test that the logic works by adding a `maxNativeZoom` to a `GridLayer`.
* [x] Add some unit test. Like loading a map at a zoom past `maxNativeZoom` and perform a check on every tile loaded.
* [x] Move the docstrings and the declaration of the `maxNativeZoom` from `TileLayer` to `GridLayer`

**Comment by perliedman** [comment]
I would guess this is also more or less exactly what @jkuebart is talking about in https://github.com/Leaflet/Leaflet.VectorGrid/pull/69, so this sounds like a win for several plugins.

**Comment by jkuebart** [comment]
Unfortunately it's not quite as simple. Setting `_tileZoom` to the clamped value makes `_update()` notice a discrepancy between `map.getZoom()` and `_tileZoom` and call `_setView()` again, indefinitely.

I'll need to take a closer look what the right value for `_tileZoom` needs to be…

**Comment by IvanSanchez** [comment]
@jkuebart How about replacing...

```
if (Math.abs(zoom - this._tileZoom) > 1) { // recursive call to _setView() }
```

with something like 

```
if (this.options.maxNativeZoom !== undefined) { zoom = Math.min(zoom, this.options.maxNativeZoom) }
if (this.options.minNativeZoom !== undefined) { zoom = Math.max(zoom, this.options.minNativeZoom) }
if (Math.abs(zoom - this._tileZoom) > 1) { // recursive call to _setView() }
```

...?

`_tileZoom` should be the zoom value for the tiles, *always*, regardless of what zoom level they are displayed. My aim is also to delete the `maxNativeZoom` logic in `TileLayer`, which is basically creating extra grid levels with double/half the size.

**Comment by jkuebart** [comment]
Yes, I worked out `_update` would probably have to clamp the value too. See #5319 for what I have. I've done some testing with TileLayer and my own GridLayer subclass, but not written tests yet.

I have to go now, will be back tomorrow.

**Comment by IvanSanchez** [comment]
@jkuebart Looking good so far! :smile:

**Comment by jkuebart** [comment]
Sorry, missed one commit, rebased now.

**Comment by IvanSanchez** [comment]
@jkuebart Please see https://github.com/jkuebart/Leaflet/pull/1 , I'm trying to add a couple of unit tests.

**Comment by IvanSanchez** [comment]
Closed by #5319 :-D

**Comment by IvanSanchez** [comment]
I almost forgot to say thanks to @jkuebart for looking at this :-)

**Comment by cmulders** [comment]
Just a heads up: this fix only works for translate3D enabled browser.

The tile/grid layer does not render properly if 3D is disabled.
Example with latest Leaflet dev version that includes this fix: https://playground-leaflet.rhcloud.com/piwu/1/edit?html,output
Try to zoom in.

The released Leaflet 1.0 does not have this problem, just swap the library in the playground.

I am aware that almost everyone has browsers with 3D capabilities, but then leaflet should drop support for non-translate3D browsers with the next release.
